src/formatting/formatter/duration.rs

```diff
@@ -0,0 +1,540 @@
+                    || (self.show_leading_units_if_zero
+                        && index >= self.min_unit_index + 1 - self.units);
+            }
+
+            if should_push {
+                v.push((index, value));
+                // We have the right number of values/units
+                if v.len() == self.units {
+                    break;
+                }
+            }
+            ms %= div;
+        }
+
+        v
+    }
+}
+
+impl Formatter for DurationFormatter {
+    fn format(&self, val: &Value, _config: &SharedConfig) -> Result<String, FormatError> {
+        match val {
+            Value::Duration(duration, _) => {
+                let mut v = self.get_time_parts(duration.as_millis());
+
+                if self.round_up {
+                    // Get the index for which unit we should round up to
+                    let i = v.last().map_or(self.min_unit_index, |&(i, _)| i);
+                    v = self.get_time_parts(duration.as_millis() + UNIT_CONVERSION_RATES[i] - 1);
+                }
+
+                let mut first_entry = true;
+                let mut result = String::new();
+                for (i, value) in v {
+                    // No separator before the first entry
+                    if !first_entry {
+                        if self.hms {
+                            // Separator between s and ms should be a '.'
+                            if i == 6 {
+                                result.push('.');
+                            } else {
+                                result.push(':');
+                            }
+                        } else {
+                            result.push(' ');
+                        }
+                    } else {
+                        first_entry = false;
+                    }
+
+                    // Pad the value
+                    let value_str = value.to_string();
+                    for _ in value_str.len()..UNIT_PAD_WIDTHS[i] {
+                        result.push_str(&self.pad_with);
+                    }
+                    result.push_str(&value_str);
+
+                    // No units in hms mode
+                    if !self.hms {
+                        if self.unit_has_space {
+                            result.push(' ');
+                        }
+                        result.push_str(UNITS[i]);
+                    }
+                }
+
+                Ok(result)
+            }
+            other => Err(FormatError::IncompatibleFormatter {
+                ty: other.type_name(),
+                fmt: "duration",
+            }),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! fmt {
+        ($name:ident) => {{
+            fmt!($name,)
+        }};
+        ($name:ident, $($key:ident : $value:tt),* $(,)?) => {
+            new_formatter(stringify!($name), &[
+                $( Arg { key: stringify!($key), val: stringify!($value) } ),*
+            ])
+        };
+    }
```

Comment:
`eng.rs` has a similar macro for tests, let's share them?

Reply:
Created #2054, will rebase after that's merged